docs/config.md
# Config

loghead looks for a config file named `config.yaml`[^1] in `./`, `~/.loghead/` and `/etc/loghead/`.

## Listeners

The services (Client logs, SSH session recording, and Client metrics aggregation) can be exposed in one of two ways
- `type: plain` - publicly available directly over the network
- `type: tsnet` - only available over tailscale as a [tsnet](https://tailscale.com/blog/tsnet-virtual-private-services) service
When a service is exposed as a tsnet service an [AuthKey](https://tailscale.com/kb/1085/auth-keys) has to be provided.

## Default config

The default config is shown below.

```yaml
log:
  level: "info"
  format: "text"

loghead:
  processors:
    # write logs to a one file per node
    filelogger:
      enabled: true
      dir: "./logs"
    # forward the logs to another logtail instance
    forward:
      enabled: false
      addr: "https://log.tailscale.io"
    # expose the metrics contained in the logs in the prometheus format
    metrics: true
    # log the node's host info to the console
    hostinfo: false
  listener:
    type: "plain" # "plain" or "tsnet"
    addr: "0.0.0.0"
    port: "5678"
    tsnet:
      authKey: ""
      controllURL: "https://controllplane.tailscale.com"
      hostname: "" # hostname of the tsnet service
      dir: "/tsnet-state/loghead" # where state is stored


ssh_recorder:
  dir: "./recordings"
  listener:
    type: "tsnet" # "plain" or "tsnet"
    addr: "0.0.0.0"
    port: "80" # `80` is used by both headscale and tailscale
    tsnet:
      authKey: ""
      controllURL: "https://controllplane.tailscale.com"
      hostname: "" # hostname of the tsnet service
      dir: "/tsnet-state/ssh_recorder" # where state is stored

node_metrics:
  listener:
    type: "plain" # "plain" or "tsnet"
    addr: "0.0.0.0"
    port: "5679"
    tsnet:
      authKey: ""
      controllURL: "https://controllplane.tailscale.com"
      hostname: "" # hostname of the tsnet service
      dir: "/tsnet-state/node_metrics" # where state is stored
```

[^1]: Only the name `config` is fixed. All formats supported by viper (JSON, YAML, TOML, HCL, envfile and Java properties config files) are supported. YAML is the recommended format.

CHANGELOG.md
# CHANGELOG

## Unreleased

- build: also build container image for linux/arm64
- build: update project description
- feat!: add defaults for tsnet listener state directories
- docs: improve docs and add docs on Client metrics

## 0.0.5 (2024-12-21)

- build(deps): update dependency file
- build: update goreleaser config
- build: set container image annotation in addition to labels
- chore: update .gitignore
- feat: implement basic support for aggregating client metrics
- refactor: rename log handling package
- refactor: extract http listener creation into func
- fix: error handling crash if binding a port failed
- build(deps): update dependencies
- feat: automatically create network logs folder for filelogger
- feat: support different collections in filelogger
- feat: process ssh session recording metadata
- feat: process ssh session recordings during the recording

## 0.0.4 (2024-02-28)

- fix: detection if ts is up
- style: bubble up and handle errors more consistently
- fix: endpoints return code 500 for fatal errors
- fix: bubble up errors that occour when starting a server
- feat: can set ts hostname and state dir via config
- feat: write ts logs to logs as level trace
- fix: log messages during ts listener startup
- build: use distroless/base-debian12 as base image
- refactor: extract ts settings into substruct
- refactor: encapsulate everything in services
- refactor!: group loghead config options
- feat: use errors that capture stacktraces
- fix: update config defaults
- docs: update docs

## 0.0.3 (2024-02-25)

- refactor: generalize HTTP listener config loading
- refactor: generalize HTTP listener creation
- feat!: add client log collection as a tsnet listener
- docs: add BSD-3-Clause license
- feat: tsnet listener waits for ts to be ready before starting

## 0.0.2 (2024-02-22)

- Add a forwarder for the client logs
- Change default port for SSH session recording to 80
- Add SSH session recording as a tsnet listener

## 0.0.1 (2024-02-20)

- Initial release


types/config.go
package types

import (
	"github.com/cockroachdb/errors"
	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
	"github.com/spf13/viper"
	"strings"
)

type Config struct {
	Log         LogConfig
	SSHRecorder SSHRecorderConfig
	Loghead     LogheadConfig
	NodeMetrics NodeMetricsConfig
}

type FileLoggerConfig struct {
	Enabled bool
	Dir     string
}

type ForwardingConfig struct {
	Enabled bool
	Addr    string
}

type LogConfig struct {
	Level  zerolog.Level
	Format string
}

type ProcessorConfig struct {
	FileLogger FileLoggerConfig
	Metrics    bool
	Hostinfo   bool
	Forward    ForwardingConfig
}

type ListenerConfig struct {
	Type string
	Addr string
	Port string
	TS   TSConfig
}

type TSConfig struct {
	AuthKey     string
	ControllURL string
	HostName    string
	Dir         string
}

type SSHRecorderConfig struct {
	Dir      string
	Listener ListenerConfig
}

type LogheadConfig struct {
	Processors ProcessorConfig
	Listener   ListenerConfig
}

type NodeMetricsConfig struct {
	Enabled  bool
	Targets  []string
	Listener ListenerConfig
}

const (
	JSONLogFormat = "json"
	TextLogFormat = "text"
)

func GetProcessorConfig() ProcessorConfig {
	return ProcessorConfig{
		FileLogger: GetFileLoggerConfig(),
		Metrics:    viper.GetBool("loghead.processors.metrics"),
		Hostinfo:   viper.GetBool("loghead.processors.hostinfo"),
		Forward:    GetForwardingConfig(),
	}
}

func GetForwardingConfig() ForwardingConfig {
	return ForwardingConfig{
		Enabled: viper.GetBool("loghead.processors.forward.enabled"),
		Addr:    viper.GetString("loghead.processors.forward.addr"),
	}
}

func GetListenerConfig(base string) ListenerConfig {
	return ListenerConfig{
		Type: viper.GetString(base + ".listener.type"),
		Addr: viper.GetString(base + ".listener.addr"),
		Port: viper.GetString(base + ".listener.port"),
		TS:   GetTSConfig(base + ".listener"),
	}
}

func GetTSConfig(base string) TSConfig {
	return TSConfig{
		AuthKey:     viper.GetString(base + ".tsnet.authKey"),
		ControllURL: viper.GetString(base + ".tsnet.controllURL"),
		HostName:    viper.GetString(base + ".tsnet.hostname"),
		Dir:         viper.GetString(base + ".tsnet.dir"),
	}
}

func GetSSHRecorderConfig() SSHRecorderConfig {
	return SSHRecorderConfig{
		Dir:      viper.GetString("ssh_recorder.dir"),
		Listener: GetListenerConfig("ssh_recorder"),
	}
}

func GetLogheadConfig() LogheadConfig {
	return LogheadConfig{
		Listener:   GetListenerConfig("loghead"),
		Processors: GetProcessorConfig(),
	}
}

func GetNodeMetricsConfig() NodeMetricsConfig {
	return NodeMetricsConfig{
		Enabled:  viper.GetBool("node_metrics.enabled"),
		Targets:  viper.GetStringSlice("node_metrics.targets"),
		Listener: GetListenerConfig("node_metrics"),
	}
}

func GetFileLoggerConfig() FileLoggerConfig {
	return FileLoggerConfig{
		Dir:     viper.GetString("loghead.processors.filelogger.dir"),
		Enabled: viper.GetBool("loghead.processors.filelogger.enabled"),
	}
}

func GetLogConfig() LogConfig {
	logLevelStr := viper.GetString("log.level")
	logLevel, err := zerolog.ParseLevel(logLevelStr)
	if err != nil {
		logLevel = zerolog.DebugLevel
	}

	logFormatOpt := viper.GetString("log.format")
	var logFormat string
	switch logFormatOpt {
	case "json":
		logFormat = JSONLogFormat
	case "text":
		logFormat = TextLogFormat
	case "":
		logFormat = TextLogFormat
	default:
		log.Error().Str("func", "GetLogConfig").
			Msgf("Could not parse log format: %s. Valid choices are 'json' or 'text'", logFormatOpt)
	}

	return LogConfig{
		Level:  logLevel,
		Format: logFormat,
	}
}

func LoadConfig() (*Config, error) {
	viper.SetConfigName("config")
	viper.AddConfigPath("/etc/loghead/")
	viper.AddConfigPath("$HOME/.loghead/")
	viper.AddConfigPath(".")

	viper.SetEnvPrefix("loghead")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	viper.AutomaticEnv()

	viper.SetDefault("log.level", "info")
	viper.SetDefault("log.format", TextLogFormat)

	viper.SetDefault("loghead.processors.filelogger.enabled", true)
	viper.SetDefault("loghead.processors.filelogger.dir", "./logs")
	viper.SetDefault("loghead.processors.forward.enabled", false)
	viper.SetDefault("loghead.processors.forward.dir", "https://log.tailscale.io")
	viper.SetDefault("loghead.processors.metrics", false)
	viper.SetDefault("loghead.processors.hostinfo", false)
	viper.SetDefault("loghead.listener.type", "plain")
	viper.SetDefault("loghead.listener.addr", "0.0.0.0")
	viper.SetDefault("loghead.listener.port", "5678")
	viper.SetDefault("loghead.listener.tsnet.controllURL", "https://controlplane.tailscale.com")
	viper.SetDefault("loghead.listener.tsnet.dir", "/tsnet-state/loghead")

	viper.SetDefault("ssh_recorder.dir", "./recordings")
	viper.SetDefault("ssh_recorder.listener.type", "tsnet")
	viper.SetDefault("ssh_recorder.listener.addr", "0.0.0.0")
	viper.SetDefault("ssh_recorder.listener.port", "80")
	viper.SetDefault("ssh_recorder.listener.tsnet.controllURL", "https://controlplane.tailscale.com")
	viper.SetDefault("ssh_recorder.listener.tsnet.dir", "/tsnet-state/ssh_recorder")

	viper.SetDefault("node_metrics.enabled", false)
	viper.SetDefault("node_metrics.targets", []string{})
	viper.SetDefault("node_metrics.listener.type", "plain")
	viper.SetDefault("node_metrics.listener.addr", "0.0.0.0")
	viper.SetDefault("node_metrics.listener.port", "5679")
	viper.SetDefault("node_metrics.listener.tsnet.controllURL", "https://controlplane.tailscale.com")
	viper.SetDefault("node_metrics.listener.tsnet.dir", "/tsnet-state/node_metrics")

	if err := viper.ReadInConfig(); err != nil {
		return nil, errors.New("Failed to read config")
	}

	var errorText string
	if (viper.GetString("ssh_recorder.listener.type") == "tsnet") && (!viper.IsSet("ssh_recorder.listener.tsnet.authKey")) {
		errorText += "Fatal config error: when using a tsnet listener, authkey must be provided\n"
	}
	if (viper.GetString("loghead.listener.type") == "tsnet") && (!viper.IsSet("loghead.listener.tsnet.authKey")) {
		errorText += "Fatal config error: when using a tsnet listener, authkey must be provided\n"
	}

	if errorText != "" {
		return nil, errors.New(strings.TrimSuffix(errorText, "\n"))
	}

	return &Config{
		Log:         GetLogConfig(),
		SSHRecorder: GetSSHRecorderConfig(),
		Loghead:     GetLogheadConfig(),
		NodeMetrics: GetNodeMetricsConfig(),
	}, nil
}
